include/boost/math/differentiation/autodiff.hpp

```diff
@@ -994,11 +994,8 @@ template<typename Func>
 fvar<RealType,Order> fvar<RealType,Order>::apply_coefficients(const size_t order, const Func& f) const
 {
   const fvar<RealType,Order> epsilon = fvar<RealType,Order>(*this).set_root(0);
-#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
-  size_t i = std::min(order, order_sum);
-#else
-  size_t i = order < order_sum ? order : order_sum;
-#endif
+  constexpr auto fvar_order_sum = fvar<RealType,Order>::order_sum;
+  size_t i = (std::min)(order, fvar_order_sum);
```

Comment:
This is a downgrade to the C++17 version. I'd rather keep the ` #ifndef BOOST_NO_CXX17_IF_CONSTEXPR`.

At some point, support for C++11 will be abandoned just as older C++ versions are not supported. When that time comes, it will be much easier if these compiler checks are in place, rather than manually changing these compromised lines back to the more succinct C++17 version.

Reply:
I would only comment (for context) that being `constexpr`, `fvar_order_sum` should be applied to `std::min` at compile time as if the argument `fvar::order_sum` were being passed directly; in other words, a no-op in terms of allocating space on the stack as would be the case for a run-time automatic variable, since the `fvar_order_sum` isn’t calculated at run-time (being a constant expression instead), and therefore inline-able for any reasonable compiler.

The ODR issue also affects C++14; it only became deprecated in C++17.

The impetus for the change is that users tend to find preprocessor separations confusing, and I saw this as a reasonable compromise which should have cause no performance degradation while improving readability (being only 1 line as opposed to several the preprocessor solution in place), while also avoiding the need for a namespace scope declaration for `fvar::order_sum`.

If you still would prefer the current code, I’ll remove this change from the PR.

Reply:
for what it's worth, I agree that this is more readable (at no performance cost) than the ifndef. Seems to me that it's better up until the point that you no longer want to support c++14 at which point it because slightly more verbose. Given the tendency (from my observation at least) for boost to support super old c++ versions I would assume that you'd keep c++14 support for a long time.

If you do decide to go with the compiler switch could you please change it from an ifndef to an ifdef. I found the double negative to be quite confusing.

Reply:
> If you still would prefer the current code, I’ll remove this change from the PR.

Yes please. My most previous [comment](https://github.com/pulver/autodiff/pull/29/files#r270627779) also applies here. Except that in this case there is indeed no performance degradation, but instead it is the loss of brevity made available to us by C++17.

Reply:
> If you do decide to go with the compiler switch could you please change it from an ifndef to an ifdef. I found the double negative to be quite confusing.

In the short term as a new user of the code I would agree with your point. In the longer term, after getting familiar with this logic, I would prefer the current version.

Conceptually (IMO) it is easier to see the higher-level code which comes closer to expressing the intent of the code block first. When reading through the code I can then mentally skip the `#else` block after understanding the meaning.